runtime/gc_include/ObjectAccessBarrierAPI.hpp

```diff
@@ -396,6 +396,54 @@ class MM_ObjectAccessBarrierAPI
 #endif /* defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER) */
 	}
 
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+	VMINLINE UDATA
+	mixedObjectGetFieldSize(J9Class *objectClass)
+	{
+		UDATA size = objectClass->totalInstanceSize;
+		if (0 <= (IDATA)objectClass->lockOffset) {
+			size -= sizeof(j9objectmonitor_t);
+		}
+		return size;
+	}
+
+	/**
+	 * Copy valueType from sourceObject to destObject
+	 * See MM_ObjectAccessBarrier::copyValue for detailed description
+	 *
+	 * @param vmThread vmthread token
+	 * @param valueClass The valueType class
+	 * @param srcObject The object being used.
+	 * @param srcOffset The offset of the field.
+	 * @param destObject The object being used.
+	 * @param destOffset The offset of the field.
+	 */
+	VMINLINE void
+	copyValue(J9VMThread *vmThread, J9Class *valueClass, j9object_t srcObject, UDATA srcOffset, j9object_t destObject, UDATA destOffset)
+	{
+#if defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER)
+		vmThread->javaVM->memoryManagerFunctions->j9gc_objaccess_copyValue(vmThread, srcObject, srcOffset, destObject, destOffset);
+#else /* defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER) */
+		/* TODO implement HW barriers */
+		if (j9gc_modron_readbar_none != _readBarrierType) {	
+			if (j9gc_modron_readbar_evacuate == _readBarrierType) {
+				vmThread->javaVM->memoryManagerFunctions->j9gc_objaccess_copyValue(vmThread, valueClass, srcObject, srcOffset, destObject, destOffset);
+			} else if (j9gc_modron_readbar_always == _readBarrierType) {
+				vmThread->javaVM->memoryManagerFunctions->j9gc_objaccess_copyValue(vmThread, valueClass, srcObject, srcOffset, destObject, destOffset);
+			}
+		} else {
+			UDATA offset = 0;
+			UDATA limit = mixedObjectGetFieldSize(valueClass);
+			while (offset < limit) {
+				*(fj9object_t *)((UDATA)destObject + offset + destOffset) = *(fj9object_t *)((UDATA)srcObject + offset + srcOffset);
+				offset += sizeof(fj9object_t);
+			}
+			postBatchStoreObject(vmThread, destObject);
+		}
+#endif /* defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER) */
+	}
+#endif /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+
```

Comment:
again, perhaps we can share it with cloneObject, having a private (it's fine to be inlined, the point is not on saving footprint, but on having more maintainable code) method with an additional argument if lockwork should be copied or not.